### ISSUE TITLE
List of Records with optional values does not work despite type annotations

### ISSUE THREAD

**Issue by xeruf** [dhall-lang/dhall-lang#1085]
I have this wonderful Dhall file:

```dhall
let Object = { name: Text, description: Optional Text }

let list : List Object = [
  {
    name = "Screwdriver",
    description = "drives screws"
  },
  {
    name = "Plate"
  }
]

in list : List Object
```

And even though I have clearly annotated the type as often as possible, I get this error:

```sh
Error : List elements should all have the same type

{ - description : …
, …
}

 8│   {
 9│     name = "Plate"
10│   }

test.dhall:8:3
```

As soon as I add the extra argument to the second object, it works without _any_ annotations. But I can't even omit the "Some" in the first one!

```dhall
[
  {
    name = "Screwdriver",
    description = Some "drives screws"
  },
  {
    name = "Plate",
    description = None Text
  }
]
```

And no, I don't want any of the ugly workarounds mentioned in #985 - I am fine with any one-time annotation/declaration, but I want the actual objects to stay clean and readable, almost like JSON. If I have to explicitly declare optional properties, they are useless to me.

**Comment by wallyqs** [comment]
Maintainers will probably chime-in but I think one of the recommended ways to make this flexible is by using `default` and letting that fill in the record:

```dhall
let Object = { 
  Type = { 
    name: Text, 
    description: Optional Text 
  },
  default = { 
    description = None Text 
  }, 
}

let list : List Object.Type = [
  Object::{
    name = "Screwdriver",
    description = Some "drives screws"
  },
  Object::{
    name = "Plate"
  },
]

in list : List Object.Type
```

Then you could have `Screwdriver` and `Plate` types for example which are different types but could be part of an union type etc... and either could then belong to the same list.

**Comment by Gabriel439** [comment]
Yeah, what @wallyqs said is the closest solution that is possible using existing language features.

I also want to throw out another idea, which is that if we had language support for importing JSON (such as suggested in https://github.com/dhall-lang/dhall-lang/issues/121), then you could literally write the list of `Object`s as a JSON value and then import that.

**Comment by xerus2000** [comment]
@Gabriel439 that sounds promising
@wallyqs thanks, but I already saw that. And it needs to many modifications. I don't want the declaration of optional record entries to be different from the non-optional ones (the `Some`) - I shouldn't need to change every object in the List if I make an existing property optional...

**Comment by wallyqs** [comment]
You also could use the `default` and only use non optional fields:

```dhall
let Object = {
  , Type = {
    , name: Text
    , description: Text
  }
  , default = {
    , name = ""
    , description = ""
  }
}

let list = [
   Object::{ 
     , name = "Screwdriver"
   },
   Object::{ 
     , name = "Plate"
     , description = "drive screws"
   }
]

in list
```

Using the optional types is really useful though since they can be used with `merge` to match into a function like:

```dhall
let Object = {
  , Type = {
    , name: Text
    , description: Optional Text
  }
  , default = {
    , name = ""
    , description = None Text
  }
}

let list : List Object.Type = [ 
   , Object::{ 
     name = "Screwdriver" 
   , description = Some "drive screws"
   }
   , Object::{ 
       name = "Plate"
   }
]

let testA = Object::{ name = "screwdriver", description = Some "example" }
let testB = Object::{ name = "plate", description = None Text }

in merge {
   , None = "Empty description"
   , Some = \(description : Text) -> description
} testA.description
```

**Comment by SiriusStarr** [comment]
> I shouldn't need to change every object in the List if I make an existing property optional...

Another solution I frequently use is to make constructors, e.g.

```dhall
let Object = { name : Text, description : Optional Text }

let withDescription
    : Text -> Text -> Object
    = \(name : Text) ->
      \(description : Text) ->
        { name, description = Some description }

let withoutDescription
    : Text -> Object
    = \(name : Text) -> { name, description = None Text }

let list
    : List Object
    = [ withDescription "Screwdriver" "drives screws"
      , withoutDescription "Plate"
      ]

in  list : List Object
```

The upside to this is if you change your type, you just edit your constructor functions rather than every thing in the list.

You can also use this pattern: 

```dhall
let Object =
      { Type =
          { name : Text, description : Optional Text, color : Optional Text }
      , default = { description = None Text, color = None Text }
      }

let withDescription =
      \(description : Text) -> { description = Some description }

let withColor = \(color : Text) -> { color = Some color }

let list
    : List Object
    = [ Object::{ name = "Screwdriver" } /\ withDescription "drives screws"
      , Object::{ name = "Plate" }
      , Object::{ name = "Cup" } /\ withDescription "a mug" /\ withColor "blue"
      ]

in  list : List Object
```

This second one is maybe my favorite, since it allows you to add new optional properties later without redoing any data and without needing `Some`'s.

**Comment by xerus2000** [comment]
@SiriusStarr yeah, kinda like that, though the syntax is overly verbose.
I would like at least the meat of the file to be readable for someone who knows JSON. It's all workarounds, can't we fix the language?

**Comment by Nadrieril** [comment]
Well, this is actually part of the design decisions of the language: every subexpression should have a single and easy-to-compute type. This is not unchangeable but runs quite deep in dhall currently

**Comment by xerus2000** [comment]
Concepts and designs are great - but I don't see this working for the majority of config files without lots of hiccups...

**Comment by Nadrieril** [comment]
I don't know about hiccups: Dhall has some learning curve, but once you got it it's quite clear where you need those `Some`s and `None`s. However I do agree that this sacrifies convenience and closeness with JSON.
I'd personally be up for considering such a feature if we could figure out a principled way of making it work. But subtyping and polymorphism are not known to be good friends, so I expect this to be complex.

**Comment by joneshf** [comment]
I've been exploring my two previous comments (https://github.com/dhall-lang/dhall-lang/issues/382#issuecomment-526641371 and https://github.com/dhall-lang/dhall-lang/issues/985#issuecomment-622390728) about optional fields for the past year and change in [`purescript-option`](https://github.com/joneshf/purescript-option).

It's written in PureScript, not Dhall. I recognize the important differences in these languages (row-polymorphism and type inference) mean that it's not something that can be directly lifted. But, maybe it can inspire a solution in Dhall?

**Comment by xerus2000** [comment]
> I don't know about hiccups: Dhall has some learning curve, but once you got it it's quite clear where you need those `Some`s and `None`s. However I do agree that this sacrifies convenience and closeness with JSON.

Yes I now know when to use what.

However, the initial usecase for me was to reduce duplication in a personal configuration file by writing it in dhall and then converting it to the original format with a quick Haskell script. That scripts has grown tremendously since it has to clearly define all types, and even the configuration file in Dhall is quite a bit longer than the original. And I haven't even added optional properties yet - at that point, I'd have to add tenths of `None`s into it...

Basically I'd like to write my configuration in a scripting language meant for configuration (like Dhall) but without all the strong typing - can I take something out of Dhall for this? Or are you aware of an alternative that might suit me better?
The issue with strong typing is that, in the end, this config is just a bunch of key-value pairs with the value being a number or a String, so there is really not much to gain in this case. I mostly considered it for deduplication reoccuring pairs and composing them with simple scripting.

**Comment by Nadrieril** [comment]
Then indeed Dhall may not be what you're looking for. Alternatives I've heard about include the recently-released [Nickel](https://www.tweag.io/blog/2020-10-22-nickel-open-sourcing), [Starlark](https://github.com/bazelbuild/starlark) and [Jsonnet](https://jsonnet.org/).

**Comment by ad-si** [comment]
Can't we support something like this?

```dhall
let Object = { 
  Type = { 
    name: Text, 
    description: Optional Text 
  },
  default = { 
    description = None Text 
  }, 
}

let list = List Object::[
  {
    name = "Screwdriver",
    description = Some "drives screws"
  },
  {
    name = "Plate"
  },
]

in list : List Object.Type
```

This feels like the natural extension to lists for me.

**Comment by Gabriel439** [comment]
@ad-si: It's possible, although the main tradeoff is that the right argument to the `::` operator would need to be treated as a special syntactic form, since if you treated it as an ordinary `List` then it would not type-check in isolation (since its elements wouldn't match).

I have proposed in another thread treating the right argument to `::` as a special syntactic form for a different reason, so that we could treat `Foo::{ a.b = c, x.y = z }` as syntactic sugar for `Foo with a.b = c with x.y = z` (so that overriding default correctly handles nesting), so it wouldn't be that odd to extend the same principle to lists.

Another issue is that it would also require treating the left-hand side of the `::` operator as a special syntactic form, too, since `List Object` would not type-check if treated as an ordinary expression (since `Object` is a record containing a type, and not a type itself).

So it's possible to do, but my inclination is still to prefer the more explicit approach where the user wraps each element of the list in `Object::`.  The main reason why is that even if `Object` happened to be a really long name in practice (e.g. `kubernetes.schemas.SomeObject`), you still have the option of creating a short-hand synonym for the schema using a `let` binding.

I think the main thing that would change my mind is if there were a significant performance improvement to being able to elide the `Object::` from each element of the list.

**Comment by basile-henry** [comment]
We could imaging a simpler sugar for defaults over lists. I don't think `Object::[ x, y, z ]` currently parses so it should be feasible to make it a sugar for `[ Object::x, Object::y, Object::z]`. We might have to think about the case where x, y, or z is not a record literal but a variable as that's not sugar that is supported, but I think it could still work with variables in the list if we skip the intermediate sugar.

So for a more complete example:

```dhall
let Object = { 
  Type = { 
    name: Text, 
    description: Optional Text 
  },
  default = { 
    description = None Text 
  }, 
}

let other = {
  name = "Banana",
}

let list = Object::[
  {
    name = "Screwdriver",
    description = Some "drives screws"
  },
  {
    name = "Plate"
  },
  other,
]

in list : List Object.Type
```
